src/script/components/asset/controls/MediaButton.tsx

```diff
@@ -0,0 +1,105 @@
+/*
+ * Wire
+ * Copyright (C) 2018 Wire Swiss GmbH
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+import cx from 'classnames';
+import ko from 'knockout';
+import AssetLoader from 'Components/asset/AssetLoader';
+import React, {useEffect, useState} from 'react';
+import type {FileAsset} from '../../../entity/message/FileAsset';
+import {AssetTransferState} from '../../../assets/AssetTransferState';
+import {noop} from 'Util/util';
+import {registerReactComponent} from 'Util/ComponentUtil';
+
+export interface MediaButtonProps {
+  asset: FileAsset;
+  cancel?: () => void;
+  large: boolean;
+  pause?: () => void;
+  play: () => void;
+  src: HTMLMediaElement;
+  transferState: AssetTransferState;
+  uploadProgress: number;
+}
+
+const MediaButton: React.FC<MediaButtonProps> = ({
+  src,
+  large,
+  asset,
+  uploadProgress,
+  transferState,
+  play,
+  pause = noop,
+  cancel = noop,
+}) => {
+  const [isPlaying, setIsPlaying] = useState<boolean>(false);
+  const onPlay = () => setIsPlaying(true);
+  const onPause = () => setIsPlaying(false);
+
+  const mediaElement = src;
+
+  useEffect(() => {
+    mediaElement.addEventListener('playing', onPlay);
+    mediaElement.addEventListener('pause', onPause);
+
+    return () => {
+      mediaElement.removeEventListener('playing', onPlay);
+      mediaElement.removeEventListener('pause', onPause);
+    };
+  }, [mediaElement]);
+
+  const isUploaded = transferState === AssetTransferState.UPLOADED;
+  const isDownloading = transferState === AssetTransferState.DOWNLOADING;
+  const isUploading = transferState === AssetTransferState.UPLOADING;
+
+  const mediaButtonPlay = (
+    <div className="media-button media-button-play icon-play" onClick={play} data-uie-name="do-play-media"></div>
+  );
+
+  const mediaButtonPause = (
+    <div className="media-button media-button-pause icon-pause" onClick={pause} data-uie-name="do-pause-media"></div>
+  );
+
+  const isDownloadingIndicator = (
+    <AssetLoader large={large} loadProgress={ko.unwrap(asset.downloadProgress)} onCancel={cancel} />
+  );
+
+  const isUploadingIndicator = <AssetLoader large={large} loadProgress={uploadProgress} onCancel={cancel} />;
+
+  return (
+    <div
+      className={cx({
+        'media-button-lg': large,
+      })}
+    >
+      {isUploaded && !isPlaying && mediaButtonPlay}
```

Comment:
Let's try to keep the jsx in the return block instead of extracting it into consts when possible.

Reply:
What's the benefit? 

To me this looks cleaner:

![image](https://user-images.githubusercontent.com/469989/107802568-cb009380-6d61-11eb-9f14-b8d37063a3e5.png)

Than this:

![image](https://user-images.githubusercontent.com/469989/107802532-bc19e100-6d61-11eb-9904-11c8de928f45.png)

Reply:
It makes sense when you are using one part multiple times. Unless that's the case its an unnecessary indirection.